### PR TITLE
Implement ATTACH of MergeTree table for s3_plain disk

### LINKED ISSUE #42153: Readonly stateless `s3` disk
**Use case**

If you simply copied MergeTree table data to s3.

**Describe the solution you'd like**

Almost the same as the `web` disk, but without the `.index` files (s3 list and head requests are used instead).

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement `ATTACH` of `MergeTree` table for `s3_plain` disk (plus some fixes for `s3_plain`)

Simple example:

```sql
-- BACKUP writes Ordinary like structure
set allow_deprecated_database_ordinary=1;
create database ordinary engine=Ordinary;

create table ordinary.test_backup_attach engine=MergeTree() order by tuple() as select * from numbers(100);
-- NOTE: name of backup ("backup") is significant.
backup table ordinary.test_backup_attach TO Disk('backup_disk_s3_plain', 'backup');

drop table ordinary.test_backup_attach;
attach table ordinary.test_backup_attach (number UInt64) engine=MergeTree() order by tuple() settings storage_policy='attach_policy_s3_plain';
```

<details>

<summary>storage configuration</summary>

```xml
<clickhouse>
    <storage_configuration>
        <disks>
            <backup_disk_s3_plain>
                <type>s3_plain</type>
                <endpoint>http://minio1:9001/root/data/disks/disk_s3_plain/</endpoint>
                <access_key_id></access_key_id>
                <secret_access_key></secret_access_key>
            </backup_disk_s3_plain>
            <attach_disk_s3_plain>
                <type>s3_plain</type>
                <!-- NOTE: /backup/ is a name of BACKUP -->
                <endpoint>http://minio1:9001/root/data/disks/disk_s3_plain/backup/</endpoint>
                <access_key_id></access_key_id>
                <secret_access_key></secret_access_key>
            </attach_disk_s3_plain>
        </disks>
        <policies>
            <attach_policy_s3_plain>
                <volumes>
                    <main>
                        <disk>attach_disk_s3_plain</disk>
                    </main>
                </volumes>
            </attach_policy_s3_plain>
        </policies>
    </storage_configuration>
    <backups>
        <allowed_disk>backup_disk_s3_plain</allowed_disk>
    </backups>
</clickhouse>
```

</details>

Follow-up for: #42232
Fixes: #42153 (cc @alexey-milovidov )
Refs: #28527 (cc @filimonov )

_P.S. Note, that this PR also includes some refactoring around metadata code, but everything is in a separate patch, hope this will help with review._

### CODE REVIEW

**Comment by azat** [comment]
>[Stress test (tsan) — Sanitizer assert (in stderr.log)](https://s3.amazonaws.com/clickhouse-test-reports/42628/06c3da0d744682885c743097ba72d928a97d1849/stress_test__tsan_.html)

- #42638 

>[Stress test (ubsan) — Cannot start clickhouse-server](https://s3.amazonaws.com/clickhouse-test-reports/42628/06c3da0d744682885c743097ba72d928a97d1849/stress_test__ubsan_.html)

```
2022.10.25 09:29:44.584985 [ 1864119 ] {} <Error> Application: DB::Exception: Part `all_1_1_0` (`data/test_19/test/all_1_1_0/`) was found on disk `local_cache_3` which is not defined in the storage policy (defined disks: s3, wrapped disks: ): Cannot attach table `test_19`.`test` from metadata file /var/lib/clickhouse/metadata/test_19/test.sql from query ATTACH TABLE test_19.test (`x` Tuple(UInt64, UInt64)) ENGINE = MergeTree ORDER BY x SETTINGS index_granularity = 8192
```

>[Stress test (debug) — Cannot start clickhouse-server](https://s3.amazonaws.com/clickhouse-test-reports/42628/06c3da0d744682885c743097ba72d928a97d1849/stress_test__debug_.html)

Looks like the culprit is `s3_check_objects_after_upload`

File: src/Disks/ObjectStorages/IObjectStorage.h [hunk 1]
```diff
@@ -66,7 +66,32 @@ class IObjectStorage
...
+
+    /// List on prefix, but it is limited only to current path, something like
+    /// on local filesystem:
+    ///
+    ///     find . -maxdepth 1 -type f
+    ///
+    /// @param path - path to list prefix in
+    /// @param children - list of child nodes
+    /// @param common_prefixes - list of common prefixes in this path, for unix
+    /// filesystem it is simply subfolders.
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/IObjectStorage.h:91]: To get list of subfolders you should use `-type d`, not `-type f`

File: src/Disks/ObjectStorages/LocalObjectStorage.cpp [hunk 2]
```diff
@@ -104,13 +104,6 @@ std::unique_ptr<WriteBufferFromFileBase> LocalObjectStorage::writeObject( /// NO
     return std::make_unique<WriteBufferFromFile>(path, buf_size, flags);
 }
 
-void LocalObjectStorage::listPrefix(const std::string & path, RelativePathsWithSize & children) const
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/LocalObjectStorage.cpp:107]: Why did you remove that?

File: src/Disks/ObjectStorages/IObjectStorage.h [hunk 3]
```diff
@@ -66,7 +66,32 @@ class IObjectStorage
...
+    virtual void listPrefixInPath(const std::string & path,
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/IObjectStorage.h:92]: Can we name those functions in more straightforward way like `findFiles` & `getDirectories()` ?

File: src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.h [hunk 4]
```diff
@@ -20,7 +21,7 @@ namespace DB
 /// It is used to allow BACKUP/RESTORE to ObjectStorage (S3/...) with the same
 /// structure as on disk MergeTree, and does not requires metadata from local
 /// disk to restore.
-class MetadataStorageFromPlainObjectStorage final : public IMetadataStorage
+class MetadataStorageFromPlainObjectStorage final : public ReadOnlyMetadataStorage
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.h:24]: Why is it inherited from ReadOnlyMetadataStorage? I mean `MetadataStorageFromPlainObjectStorage` doesn't look like it should be readonly. Probably a comment is needed.

File: tests/integration/test_attach_backup_from_s3_plain/test.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    create table ordinary.test_backup_attach engine=MergeTree() order by tuple() as select * from numbers(100);
+    -- NOTE: name of backup ("backup") is significant.
+    backup table ordinary.test_backup_attach TO Disk('backup_disk_s3_plain', 'backup');
+
+    drop table ordinary.test_backup_attach;
+    attach table ordinary.test_backup_attach (number UInt64) engine=MergeTree() order by tuple() settings storage_policy='attach_policy_s3_plain';
+    """
+    )
+
+    assert int(node.query("select count() from ordinary.test_backup_attach")) == 100
```
Comment (vitlibar) [inline tests/integration/test_attach_backup_from_s3_plain/test.py:40]: What will happen if ClickHouse tries to merge parts of `test_backup_attach` ?

File: src/Disks/ObjectStorages/IObjectStorage.h (hunk 1 above)
Comment (azat) [reply src/Disks/ObjectStorages/IObjectStorage.h:91]: Actually it is even `find . -maxdepth 1`, folders goes to `common_prefixes` and files to `children`

File: src/Disks/ObjectStorages/LocalObjectStorage.cpp (hunk 2 above)
Comment (azat) [reply src/Disks/ObjectStorages/LocalObjectStorage.cpp:107]: It hadn't been compatible with the only user of this API - restore for `send_metadata`, for details see 333ac44c1483548e8e26daf050fcb98bf1cb478f

File: src/Disks/ObjectStorages/IObjectStorage.h (hunk 3 above)
Comment (azat) [reply src/Disks/ObjectStorages/IObjectStorage.h:92]: I guess, but I did not want to mix files/folders with S3/GCS/Azure/e.t.c. since they do not have such terms.
And also it is not `findAllFiles` but, `findAllFilesAndDirectories`.
I also thought about adding this is an argument, but came to conclusion that it will be worse.

File: src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.h (hunk 4 above)
Comment (azat) [reply src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorage.h:24]: Yes, it is not completely readonly, but it does not have any extended metadata support (chown/mtime/stat/...) and `ReadOnlyMetadataStorage` throws `NOT_IMPLEMENTED` exception for this API.

I thought about the name for `ReadOnlyMetadataStorage`, and everything else looks more complex and less cleaner.

Maybe you have any other ideas?

>Probably a comment is needed

Ok (at first I had a comment for ReadOnlyMetadataTransaction/ReadOnlyMetadataStorage but eventually decided that it will only confuse reader)

File: tests/integration/test_attach_backup_from_s3_plain/test.py (hunk 5 above)
Comment (azat) [reply tests/integration/test_attach_backup_from_s3_plain/test.py:40]: It will throw some error, since it does not support mtime/move/...

Comment (vitlibar) [reply tests/integration/test_attach_backup_from_s3_plain/test.py:40]: But merges will be assigned automatically, won't they? So it seems after a while such MergeTree table will spam in logs that it cannot do merging.

File: src/Disks/ObjectStorages/IObjectStorage.h (hunk 3 above)
Comment (vitlibar) [reply src/Disks/ObjectStorages/IObjectStorage.h:92]: I mean `listPrefix()` can be named `findAllFiles` (it kind of finds all files in a directory recursively) and `listPrefixInPath` can be named `getDirectoryContents()` (it lists both files and subdirectories in a directory).

File: src/Disks/ObjectStorages/LocalObjectStorage.cpp (hunk 2 above)
Comment (vitlibar) [reply src/Disks/ObjectStorages/LocalObjectStorage.cpp:107]: Can you explain a bit more why it had been incompatible? I mean for `LocalObjectStorage` the implementation of those functions seems like it must be quite simple and standard.

**Comment by vitlibar** [comment]
What do you thing can we implement something like `hdfs_plain` disk in the future?

File: src/Disks/ObjectStorages/IObjectStorage.h (hunk 3 above)
Comment (azat) [reply src/Disks/ObjectStorages/IObjectStorage.h:92]: Ok, will rename. I don't have strong feeling about this namings.

**Comment by azat** [comment]
>What do you thing can we implement something like hdfs_plain disk in future?

Sure, it will be very simple.

File: src/Disks/ObjectStorages/LocalObjectStorage.cpp (hunk 2 above)
Comment (azat) [reply src/Disks/ObjectStorages/LocalObjectStorage.cpp:107]: >Can you explain a bit more why it had been incompatible?

```
void LocalObjectStorage::listPrefix(const std::string & path, RelativePathsWithSize & children) const
{
    fs::directory_iterator end_it;
    for (auto it = fs::directory_iterator(path); it != end_it; ++it)
        children.emplace_back(it->path().filename(), it->file_size());
}
```

As you can see it does not uses `directory_recursive_iterator` here, and proper `listPrefix` should:
a) be recursive
b) return only files

Both of this wasn't satisfied.
Plus they are not used in the code anyway, since you cannot store metadata (`send_metadata`) in the local disk, and here is the piece of code that creates them:

```
DiskObjectStoragePtr DiskLocal::createDiskObjectStorage()
{
    auto object_storage = std::make_shared<LocalObjectStorage>();
    auto metadata_storage = std::make_shared<FakeMetadataStorageFromDisk>(
        /* metadata_storage */std::static_pointer_cast<DiskLocal>(shared_from_this()),
        object_storage,
        /* object_storage_root_path */getPath());

    return std::make_shared<DiskObjectStorage>(
        getName(),
        disk_path,
        "Local",
        metadata_storage,
        object_storage,
        false, // <-- here, this is send_metadata=false
        /* threadpool_size */16
    );
}
```

> I mean for LocalObjectStorage the implementation of those functions seems like it must be quite simple and standard.

Yes, it can, but I prefer to remove them to avoid ambiguity, anyway this code cannot be tested now.

File: tests/integration/test_attach_backup_from_s3_plain/test.py (hunk 5 above)
Comment (azat) [reply tests/integration/test_attach_backup_from_s3_plain/test.py:40]: Yes and this is like with `web` disk right now, eventually you will get some errors in the log, for proper support some "readonly" property of tables should be implemented.

And I completely agree with you that this should be done.

P.S. moreover there are random SIGSEGVs in web disk when you will try to do this and not only this.

Comment (azat) [reply tests/integration/test_attach_backup_from_s3_plain/test.py:40]: So to clarify, I'm suggesting fixing this separately (since there are other places to fix, like you should not do OPTIMIZE for read-only tables, and isReadOnly should be extended)

File: src/Disks/ObjectStorages/LocalObjectStorage.cpp (hunk 2 above)
Comment (vitlibar) [reply src/Disks/ObjectStorages/LocalObjectStorage.cpp:107]: Yeah, I got it, so it was not incompatible, it was rather wrong.

I'd implement those functions for LocalObjectStorage anyway even if they probably won't be used right now. Just in the sake of consistency and kind of an alternative way to document those too functions better (because operations with filesystem can be considered as basic and simplest). Though I don't insist.

Comment (azat) [reply src/Disks/ObjectStorages/LocalObjectStorage.cpp:107]: My point is that implementing them w/o tests and w/o actual usage in the code is useless and can be even worse, since they can not work correctly and someone may rely on this.
Anyway I think that even if this should be done, this can be done separately.

File: tests/integration/test_attach_backup_from_s3_plain/test.py (hunk 5 above)
Comment (vitlibar) [reply tests/integration/test_attach_backup_from_s3_plain/test.py:40]: Yes I agree, the current implementation must be useful, some issues can be fixed later.

File: src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp [hunk 6]
```diff
@@ -141,7 +141,7 @@ std::unique_ptr<WriteBufferFromFileBase> AzureObjectStorage::writeObject( /// NO
     return std::make_unique<WriteIndirectBufferFromRemoteFS>(std::move(buffer), std::move(finalize_callback), object.absolute_path);
 }
 
-void AzureObjectStorage::listPrefix(const std::string & path, RelativePathsWithSize & children) const
+void AzureObjectStorage::findAllFiles(const std::string & path, RelativePathsWithSize & children) const
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:144]: Is this implementation correct?

File: src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp [hunk 7]
```diff
@@ -101,18 +101,6 @@ std::unique_ptr<WriteBufferFromFileBase> HDFSObjectStorage::writeObject( /// NOL
...
-{
-    const size_t begin_of_path = path.find('/', path.find("//") + 2);
-    int32_t num_entries;
-    auto * files_list = hdfsListDirectory(hdfs_fs.get(), path.substr(begin_of_path).c_str(), &num_entries);
-    if (num_entries == -1)
-        throw Exception(ErrorCodes::HDFS_ERROR, "HDFSDelete failed with path: " + path);
-
-    for (int32_t i = 0; i < num_entries; ++i)
-        children.emplace_back(files_list[i].mName, files_list[i].mSize);
-}
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp:114]: Was this implementation wrong?

File: src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp (hunk 6 above)
Comment (azat) [reply src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:144]: I think so, but I have not checked.

Comment (azat) [reply src/Disks/ObjectStorages/AzureBlobStorage/AzureObjectStorage.cpp:144]: I leaved it since it can be used for `send_metadata`

File: src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp (hunk 7 above)
Comment (azat) [reply src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp:114]: Yes, since this is analog of `ls -l` AFAIR, and it is not recursive.

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+namespace ErrorCodes
+{
+    extern const int NOT_IMPLEMENTED;
+}
+
+class ReadOnlyMetadataStorage;
+
+/// Readonly, throws NOT_IMPLEMENTED error.
+/// Can be used to add limited read-only support of MergeTree.
+class ReadOnlyMetadataTransaction : public IMetadataTransaction
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:18]: Probably bad naming. "Read-only transaction" looks strange. I mean a transaction is always changing something in a batch, how can it be readonly?

File: src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp (hunk 7 above)
Comment (vitlibar) [reply src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp:114]: So you've deleted it for now but in order to implement `hdfs_plain` we'll need to add it back, right? In this case I suppose it would be better to keep the old implementation but probable comment it with `#if 0` and a comment.

Comment (azat) [reply src/Disks/ObjectStorages/HDFS/HDFSObjectStorage.cpp:114]: >So you've deleted it for now but in order to implement hdfs_plain we'll need to add it back, right? In this case I suppose it would be better to keep the old implementation but probable comment it with #if 0 and a comment.

For it we need both, `findAllFiles` and `getDirectoryContents`, so it should be reimplemented anyway.

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        /// Required by MergeTree
+        return {};
+    }
+    time_t getLastChanged(const std::string & /* path */) const override
+    {
+        return {};
+    }
+
+    struct stat stat(const String & /* path */) const override
```
[the remark this answers is not in the document]
Comment (azat) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:132]: I prefer to throw NOT_IMPLEMENTED for every API that is not required for now, since who knows who can rely on this...

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h (hunk 8 above)
Comment (azat) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:18]: Yes, I thought about for awhile but did not managed to came up with a better alternative. Do you have some ideas?

P.S. I thought about using Limited/Surrogate/... but all of these also sound strange and does not clarify it a lot.

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:132]: It's perfectly fine to throw NOT_IMPLEMENTED when you don't have an implementation. It just looks confusing because for example `getLastChanged()` returns 0 and `readFileToString()` throws an exception. Why is that difference?

[the remark this answers is not in the document]
Comment (azat) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:132]: This was based on `MetadataStorageFromStaticFilesWebServer`, and for it `getLastChanged` returns `0`.
But AFAICS MergeTree code relies on it, though for these type of disks (s3_plain and web) it should not work anyway and returning 0 is not good since this breaks the behaviour.

[the remark this answers is not in the document]
Comment (azat) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:132]: While for example getLastModified required by MergeTree to work, even for read-only.

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h (hunk 8 above)
Comment (vitlibar) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:18]: Maybe `FakeMetadataTransactionForReadOnlyStorage` will be more clear, what do you think?

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (vitlibar) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:136]: The main problem with those two classes `ReadOnlyMetadataStorage` and `ReadOnlyMetadataTransaction` is that they're not consistent with themselves or at least with what seems from their names. I'd prefer them to be like this:

```
class ReadOnlyMetadataTransaction
{
   public:
      // all state-changing functions throw exception "that work because object storage is readonly"
      // maybe except commit (because there is always nothing to commit for this kind of transaction
};

class ReadOnlyMetadataStorage
{
   public:
      // all state-changing functions throw exception "that doesn't work because object storage is readonly"

      // state-reading functions are either left pure virtual or (if it's really necessary) throw exception 
      // but with a different text (just "that is not implemented", not like the message we throw for state-changing functions)

      // and if you want some state-reading or state-changing function to be something else
      // it can be done by overriding functions in inherited class
};
```

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h (hunk 8 above)
Comment (azat) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:18]: a) it is toooo long
b) there is already `FakeMetadataStorageFromDisk` 
c) it does not give any benefit since you do not need transactions for readonly storage

Ok, let's just throw NOT_IMPLEMENTED from all methods.

Comment (vitlibar) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:18]: ok

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        throwNotImplemented();
+    }
+
+    bool supportsChmod() const override { return false; }
+    bool supportsStat() const override { return false; }
+
+private:
+    [[noreturn]] static void throwNotImplemented()
+    {
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Operation is ont implemented");
```
Comment (vitlibar) [inline src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:142]: **not** implemented

Could you please also make `getLastModified` and `getHardlinkCount` follow the same pattern (I mean those function should rather call `throwNotImplemented()` because what they do right now has nothing to do with readonlyness.

**Review by vitlibar — APPROVED** [review]

File: src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h (hunk 10 above)
Comment (azat) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:142]: And now this becomes looks more and more like `NotImplementedMetadataStorage` then `ReadOnlyMetadataStorage`.
So I'm suggesting remove this classes completely and instead use throw over pure-virtual for the interface classes then.

Comment (vitlibar) [reply src/Disks/ObjectStorages/ReadOnlyMetadataStorage.h:142]: For transaction probably it's better to keep the readyonly class because it throws a different exception (however it's quite a weak reason).

**Comment by azat** [comment]
Failures does not looks related:

>[Stress test (tsan) — Sanitizer assert (in stderr.log)](https://s3.amazonaws.com/clickhouse-test-reports/42628/c93262170df0c0fb09ccbb56f5f73b7a41bc58fc/stress_test__tsan_.html)

- #42640 

>[Integration tests (asan) [1/3] — fail: 1, passed: 666, flaky: 5](https://s3.amazonaws.com/clickhouse-test-reports/42628/c93262170df0c0fb09ccbb56f5f73b7a41bc58fc/integration_tests__asan__[1/3].html)

- `test_distributed_ddl/test.py::test_replicated_without_arguments[configs_secure]`

```
2022-10-31 14:28:37 [ 331 ] DEBUG : Retry 20 got exception Client failed! Return code: 57, stderr: Received exception from server (version 22.11.1):
Code: 57. DB::Exception: Received from 172.16.15.6:9000. DB::Exception: There was an error on [ch4:9440]: Code: 57. DB::Exception: Table test_ordinary.rmt already exists. (TABLE_ALREADY_EXISTS) (version 22.11.1.1). Stack trace:
```

File: tests/integration/test_attach_backup_from_s3_plain/test.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    -- BACKUP writes Ordinary like structure
+    set allow_deprecated_database_ordinary=1;
+    create database ordinary engine=Ordinary;
+
+    create table ordinary.test_backup_attach engine=MergeTree() order by tuple() as select * from numbers(100);
+    -- NOTE: name of backup ("backup") is significant.
+    backup table ordinary.test_backup_attach TO Disk('backup_disk_s3_plain', 'backup');
+
+    drop table ordinary.test_backup_attach;
+    attach table ordinary.test_backup_attach (number UInt64) engine=MergeTree() order by tuple() settings storage_policy='attach_policy_s3_plain';
```
Comment (azat) [inline tests/integration/test_attach_backup_from_s3_plain/test.py:36]: And just a note, this will not work for the case with multiple parts, due to files deduplication, but you can workaround this with the following code (but note, that it touches the backup itself, so please, be aware and do not use it on real backups):

```shell
# NOTE: here I use gsutil over s3cmd, since I tested with GCP
gsutil cat gs://path/to/backup_name/.backup > .backup
# NOTE: compose is required to overcome the following issue:
# BadRequestException: 400 Rewriting objects created via Multipart Upload is not implemented yet. As a workaround, you can use compose to overwrite the object (by specifying path with the problem as both the source and output of compose) prior to rewrite.                                               
cat .backup | xq -r '.config.contents.file[] | select(.data_file != null) | [.name, .data_file] | @tsv' | awk -vroot=gs://path/to/backup_name/ '{printf("gsutil cp %s %s || { gsutil compose %s && gsutil cp %s %s; }\n", root $2, root $1, root $2, root $2, root $1)}' > /tmp/backup-data_file-restore.sh
time bash -x backup-data_file-restore.sh |& ts | tee backup-data_file-restore.log
```
